### DISCUSSION TITLE
RAND_status() doesn't return 1 after "Making all applications use the FIPS module by default".

### DISCUSSION THREAD

**Discussion by Syllinia** [openssl/openssl#24665] [Q&A]
I'm testing FIPS on OpenSSL 3.1.4 and make changes as per "Making all applications use the FIPS module by default" in the below document.
https://www.openssl.org/docs/man3.0/man7/fips_module.html

On our system after enabling fips mode, the system will restart and "all applications use the FIPS module by default". Then the sshd process can't be started because RAND_status() doesn't return 1.

What's more, I make the below change but still doesn't work. The new code is enclosed in the macro WITH_OPENSSL.
`#ifdef WITH_OPENSSL
        int ret = RAND_poll();
        logit("<%s:%d> RAND_poll returns %d\n", __func__, __LINE__, ret);
#endif

        if (RAND_status() != 1)
                fatal("PRNG is not seeded");
`

![image](https://github.com/openssl/openssl/assets/7693528/17aebaac-9000-496f-9fcf-eb7679ba5fbb)
In the above screenshot, we can see RAND_poll() returns 1 but RAND_status() still returns non 1 which indicates the random generator is still not seeded with **enough** data.

Seems I encounter the below issue. Any suggestions? Thanks in advance.
![image](https://github.com/openssl/openssl/assets/7693528/7db07378-6764-4e39-bcc4-818dc747d135)
https://unix.stackexchange.com/questions/324209/when-to-use-dev-random-vs-dev-urandom

**Comment by paulidale** [comment]
You have likely found the cause.  Seeding takes time and on some platforms it can take a long time.

**Comment by Syllinia** [comment]
@paulidale  Thanks. Seems 3 seconds is still not enough.

![image](https://github.com/openssl/openssl/assets/7693528/ea91ecdf-ee11-4211-aeba-ad73d51ed0b4)


`
        int times = 0;
        int ret = 0;

        while ((RAND_status() != 1) && (times < 3000)) {
#ifdef WITH_OPENSSL
                ret = RAND_poll();
#endif
                usleep(1000);
                ++times;
                if (0 == times % 1000) {
                        logit("<%s:%d> RAND_status still returns 0 and trapped in the loop. times = %d", __func__, __LINE__, times);
                }
        }

        if (RAND_status() != 1) {
                fatal("PRNG is not seeded");
        } else {
                logit("<%s:%d> RAND_status returns 1");
        }
`

**Reply by paulidale** [reply]
Getting good entropy can be very difficult for some devices.  Embedded network devices are the worst I've seen.  They essentially rely on network activity as their only entropy source and, during testing, there often isn't much going on.

If you have any kind of built in RNG, it would be worth using it as a seed source.

**Reply by Syllinia** [reply]
Thanks so much for your advice.

`        while ((RAND_status() != 1)) {
#ifdef WITH_OPENSSL
                ret = RAND_poll();
#endif
                usleep(1000);
                ++times;
                if (0 == times % 1000) {
                        logit("<%s:%d> RAND_status still returns 0 and trapped in the loop. times = %d", __func__, __LINE__, times);
                }
        }

        if (RAND_status() != 1) {
                fatal("PRNG is not seeded");
        } else {
                logit("<%s:%d> RAND_status returns 1", __func__, __LINE__);
        }
`
![image](https://github.com/openssl/openssl/assets/7693528/25df4580-88a2-4287-b1fd-fd0b4065b512)

**Comment by t8m** [comment]
Do you have both fips and base providers enabled?

**Reply by Syllinia** [reply]
Yes. I only enabled these 2 providers.

**Comment by Syllinia** [comment]
I added --with-rand-seed=devrandom in the Configure options and remove the below codes in Openssl 3.1.4. However, the issue persists.
`--- openssl-openssl-3.1.4-quic1/include/crypto/rand.h   2023-10-26 07:13:18.000000000 -0700
+++ openssl-openssl-3.1.4-quic1_urandom/include/crypto/rand.h   2024-06-18 11:37:33.555879300 -0700
@@ -42,9 +42,6 @@
  */
 # define DEVRANDOM "/dev/urandom", "/dev/random", "/dev/hwrng", "/dev/srandom"
 # if defined(__linux) && !defined(__ANDROID__)
-#  ifndef DEVRANDOM_WAIT
-#   define DEVRANDOM_WAIT   "/dev/random"
-#  endif
 /*
  * Linux kernels 4.8 and later changes how their random device works and there
  * is no reliable way to tell that /dev/urandom has been seeded -- getentropy(2)
`

There're output of the below commands and each execution generates different data which indicates that /dev/random and /dev/urandom has already been written with random data.
cat /dev/urandom | base64 | head -c 1024
cat /dev/random | base64 | head -c 1024

![image](https://github.com/openssl/openssl/assets/7693528/91199479-99d4-49cb-b306-91ff4676591f)

**Reply by paulidale** [reply]
That error message indicates that it's not getting as far as collecting entropy.  It's trying to create a DRBG and it isn't able to find one.  Most likely a configuration problem.  What's the output from `openssl list -providers` ?

**Reply by Syllinia** [reply]
Thanks for your suggestion. Let me check the configuration.

![image](https://github.com/openssl/openssl/assets/7693528/4805b5e1-81a1-4e8a-bb50-9aaff6122e39)

**Reply by paulidale** [reply]
The FIPS provider isn't loading.  That's the problem.  You should see something akin to this:
```
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.4.0
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.4.0
    status: active
```

**Reply by Syllinia** [reply]
![image](https://github.com/openssl/openssl/assets/7693528/f57b5ba2-8dd3-4d97-820a-78c23ee0c3e2)

![image](https://github.com/openssl/openssl/assets/7693528/f6334c8e-ed75-4c0b-bf32-f39a5caa270b)

![image](https://github.com/openssl/openssl/assets/7693528/f33b9f67-4608-4963-b1ea-327c12ac0f91)

**Comment by Syllinia** [comment]
I use the below code to install fips module and /usr/ssl/fipsmodule.cnf is generated.
system("openssl fipsinstall -out /usr/ssl/fipsmodule.cnf -module /lib/fips.so &> /var/log/fips_install");

/var/log/fips_install seems good.
![image](https://github.com/openssl/openssl/assets/7693528/29ecaeb6-dcae-415f-a533-457759dda6a2)
![image](https://github.com/openssl/openssl/assets/7693528/3b41a07d-ce8f-4acf-9d92-78a9d517c83f)

**Reply by t8m** [reply]
3.1.4+quic build is not something we provide. Could you please try first with pristine 3.1.4 sources?

**Reply by Syllinia** [reply]
@t8m @paulidale On Ubuntu 20.04, 3.1.4+quic works well and fipsinstall can work.

./openssl fipsinstall -out /home/inst/fipsmodule.cnf -module /home/inst/lib64/ossl-modules/fips.so

./openssl list -providers
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.1.4
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.1.4
    status: active

./openssl rand -hex 16
5add90e55f0e02f5c970104b98f350a3

./openssl version -d
OPENSSLDIR: "/usr/local/ssl"

./openssl version -v
OpenSSL 3.1.4+quic 24 Oct 2023 (Library: OpenSSL 3.1.4+quic 24 Oct 2023)

cat /usr/local/ssl/openssl.cnf
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

.include /home/inst/fipsmodule.cnf

[openssl_init]
providers = provider_sect
# You MUST uncomment the following line to operate in a FIPS approved manner,
# It is commented out here purely for testing purposes.
alg_section = evp_properties

[evp_properties]
default_properties = "fips=yes"

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1

cat /home/inst/fipsmodule.cnf
[fips_sect]
activate = 1
install-version = 1
conditional-errors = 1
security-checks = 1
tls1-prf-ems-check = 0
drbg-no-trunc-md = 0
module-mac = D9:43:7A:63:6F:CB:A3:65:FF:8F:7A:94:2A:36:C6:E8:45:57:B0:89:C6:6E:FD:1F:36:34:8E:55:B6:57:16:5A

However, in my embedded system using Linux kernel, it doesn't work. Could you please give me some advice? Thanks in advance.
![image](https://github.com/openssl/openssl/assets/7693528/01120c52-8ec2-4e1e-a018-f7c2cf1b1b8e)

![image](https://github.com/openssl/openssl/assets/7693528/96182c14-d572-4537-b866-54c6ae66052d)

![image](https://github.com/openssl/openssl/assets/7693528/df36a1c6-81e0-490f-89ed-a62ea23153b4)

![image](https://github.com/openssl/openssl/assets/7693528/7081c4c5-198c-4cf4-97e0-e7fbe7f79b92)
![image](https://github.com/openssl/openssl/assets/7693528/d018dc2f-4157-4d5c-bf71-b594c6b75004)
![image](https://github.com/openssl/openssl/assets/7693528/ef627c8f-12f8-404e-87b3-4b1045ba928e)

**Comment by paulidale** [comment]
@t8m's point was that we do not support modified versions of OpenSSL which 3.1.4+quic is.
I suggest you contact the maintainer of 3.1.4+quic and ask for their assistance.

**Reply by Syllinia** [reply]
Actually, 3.1.4 + quic seems make minor changes to 3.1.4 which may not affect fips provider loading.

**Reply by Syllinia** [reply]
fips.so and libcrypto.so.3 are in the same directory. After moving fips.so to ossl-modules subdirectory, I can get both fips and base providers in output of "openssl list -proivders".
